### DISCUSSION TITLE
Parsing of instructions area to improve the turbowarp button addon

### DISCUSSION THREAD

**Discussion by sfederici** [ScratchAddons/ScratchAddons#5371] [Ideas] [answered]
Often, when a Scratch project is better in TW, project creators put a TW link in their instructions area with all the necessary parameters, such as for example

https://turbowarp.org/765401317?fps=60&stuck&offscreen&limitless

that you can see in the instructions area of project https://scratch.mit.edu/projects/765401317.

My suggestion is:

- parsing the instructions area, in order to extract the http://turbowarp.org/etc link, if any
- changing the link in the TW button addon so that it will use the extracted link instead of the general http://turbowarp.org/ link

It is true that sometimes Scratchers put TW links only in a list in their projects (that is visible when the project is started). This is useful for users that do now use ScratchAddons, but I'm sure that, when this addon (or this improvement of the TW button addon) is out, they will start adding the TW link to the instructions area too.

**Comment by BroJac5246** [comment]
Great idea!

I'm just going to note that we should make sure that the parsed link uses https.

**Reply by WorldLanguages** [reply]
And I'm not sure if it would be a good idea to support every possible parameter.

**Reply by retronbv** [reply]
with the new turbowarp feature (unsandboxed extensions) that could definitely be dangerous
garbo has put in security measures to only allow code from `extensions.turbowarp.org` (needs his approval to be added) or `localhost:8000` (for development)

https://github.com/TurboWarp/scratch-gui/blob/213b6fdb958d81686e3c63b7205ccefedd7beda4/src/containers/tw-security-manager.jsx#L8

**Reply by sfederici** [reply]
> with the new turbowarp feature (unsandboxed extensions) that could definitely be dangerous

I understand that. Usually the link in the instructions is just to run faster a project that runs in Scratch. So, I guess removing all extension-related parameters would be the good thing to do

**Reply by retronbv** [reply]
Yeah, turbowarp has safety for the unsandboxed extensions anyway, I just wanted to bring it up.

**Reply by WorldLanguages** [reply]
> So, I guess removing all extension-related parameters would be the good thing to do

We would probably use a whitelist approach, because we don't know what parameters TurboWarp will support in the future which could be dangerous.

**Comment by DNin01** [comment]
The "Store settings in project" button can achieve this behavior much more effectively and safely.

**Comment by sfederici** [comment]
I don't know about this addon. If I understand it correctly, this allows you to save your settings in a turbowarp project? If so, how will this allow you to run a project with the best TW settings when clicking the TW button in the project page?

**Comment by DNin01** [comment]
> I don't know about this addon. If I understand it correctly, this allows you to save your settings in a turbowarp project? If so, how will this allow you to run a project with the best TW settings when clicking the TW button in the project page? 

No, it's a TurboWarp feature. You can find in in Advanced on the menu bar at turbowarp.org.

The cool thing is, if you open a project containing saved settings on TurboWarp (from the normal URL), they will automatically be applied (you will be redirected to the URL with those settings), so Scratch Addons doesn't really have anything to do here.

**Reply by WorldLanguages** [reply]
@sfederici please reply below the comments!

**Reply by sfederici** [reply]
What do you mean? Did I reply in the wrong place? I see my reply right after @DNin01's comment

**Reply by sfederici** [reply]
> The cool thing is, if you open a project containing saved settings on TurboWarp (from the normal URL)

So, the settings are kept in the project uploaded to the Scratch website and are automatically applied when you click the TW button? If so, this is cool. But the problem is that not everyone suggesting specific settings will save the settings in the project, right? So the TW button won't apply the necessary settings. Or am I wrong?

**Reply by DNin01** [reply]
That's right. Not everyone will take advantage of this feature, so technically this proposed addon feature still has a use.

**Comment by sfederici** [comment, marked as answer]
This addon is not needed anymore, as the "turbowarp button" addon just makes it useless. Indeed, when you click the turbowarp button, you see the project in the turbowarp website and links in the instruction area are active as requested. So, I'm just going to close this.
